### ISSUE TITLE
File Manager. Dialog Confirm overwrite does not behave as expected.

### ISSUE THREAD

**Issue by grummbeer** [rustdesk/rustdesk#3508]
### Bug Description

The confirmation dialog for overwriting an existing file does not work properly.

1. sometimes it does not appear at all, files are silently skipped.
2. sometimes it appears only in one direction. Forgot which one it was, can't reproduce it at the moment because of "1."
3. when selecting multiple files, e.g. "select all", many dialogs are opened, but not as many as files were selected.
4. "Do this for all conflicts" does not work, behaves the same as "close this dialog".

### How to Reproduce

Open File Manger, select more than 5 files and click "receive" or "send", repeat it, play around.

### Expected Behavior

1. Dialog always displayed when a file already exists in goal.
2. "Do this for all conflicts" do so.

### Operating system(s) on local side and remote side

Linux > Android

### RustDesk Version(s) on local side and remote side

source

### Screenshots

Selected files are more than 5.

![confirm-overwrite-buggy](https://user-images.githubusercontent.com/67791701/222898488-3c7bc086-1138-48ca-9e6d-e40c2ae70f56.gif)

Dialogs disappear after a few clicks, but selected files are more.

Wanted to create a new video, but can't reproduce this anymore, instead i experienced that the files are skipped silently.

![confirm-overwrite-skipped-automaticlly](https://user-images.githubusercontent.com/67791701/222898541-bacff09b-c482-4a66-8938-2625089fa24a.gif)

**Comment by rustdesk** [comment]
@Heap-Hop

**Comment by rustdesk** [comment]
@grummbeer could you confirm if this is fixed?

**Comment by rustdesk** [comment]
> @grummbeer could you confirm if this is fixed?

Ignore this please, not fix this yet.

**Comment by Heap-Hop** [comment]
> 1. sometimes it does not appear at all, files are silently skipped.
> 2. sometimes it appears only in one direction. Forgot which one it was, can't reproduce it at the moment because of "1."
> 3. when selecting multiple files, e.g. "select all", many dialogs are opened, but not as many as files were selected.
> 4. "Do this for all conflicts" does not work, behaves the same as "close this dialog".

1,2,3 are expected. The current rule for handling same files is that **if the file name, creation date, and modification date all are the same at both ends, it will be skipped by default**. In other cases, the dialog will be showed.
 
4 is a bug. Need to fix this first. 
Also, `many dialogs are opened` (in 3.) which is the same problem.
The confirm dialog of multiple files need to be processed one by one, instead of showing all the dialogs at the same time that make "Do this for all conflicts" don't work.

**Comment by rustdesk** [comment]
@grummbeer Could you try this out? need to update on both sides, nightly build not finished yet.

**Comment by grummbeer** [comment]
cool! much better now.

🗹 double click to file now unselect the file instead of complaining "its not a dir"
🗹 no more multiple dialogs
🗹 "do this for all conflicts" do so 
🗹 nice! "this file is identical to peer's one"

❓"Receive" shows the dialog on every request, but "send" still skips the file silently. Or do i miss something?

**Comment by rustdesk** [comment]
> ❓"Receive" shows the dialog on every request, but "send" still skips the file silently. Or do i miss something?

@Kingtous

**Comment by rustdesk** [comment]
@grummbeer both sides are the new build?

**Comment by grummbeer** [comment]
yes, 40 min ago

**Comment by grummbeer** [comment]
"Home"- Button always depend on host.

Linux > Windows (click "Home") 

> flutter: Failed to openDirectory /home/stefan: 

"/home/stefan" is home of the left side (Linux)

Same vice versa with "Windows > Linux".

I guess both "sides" are always treated as OS of the left side.

**Comment by rustdesk** [comment]
@grummbeer Thanks for your detailed testing.

**Comment by grummbeer** [comment]
@Kingtous … strange, i am on master, clean and fresh build on working machine and all VMs and apk. As I routinely do, but will check this again.

**Comment by rustdesk** [comment]
@Kingtous follow up these not-clickable issue please.

**Comment by Kingtous** [comment]
> @Kingtous follow up these not-clickable issue also please.

ok i will follow it.

**Comment by grummbeer** [comment]
I can confirm "Linux > Android" behave as expected, as shown in the video. 

Note: There was a [bug with the paths when connected "linux > windows"](https://github.com/rustdesk/rustdesk/pull/3498#issuecomment-1454169188). It is fixed, but maybe has something to with this new "paths not clickable".

Sorry for being pedantic, but "Linux > Linux" still fail … could you confirm it is working?

**Comment by grummbeer** [comment]
Windows (VM) > Linux (VM; Ubuntu2204)
---

### 🗹 root level / clickable entries

Yesterday: 

![windows-linux-root-path](https://user-images.githubusercontent.com/67791701/225321075-580ba83d-5fb3-43d4-8466-02712da9a01d.png)

Now: 

![windows-linux-root-level-1-ok](https://user-images.githubusercontent.com/67791701/225555194-dda39212-21ba-49ba-839b-ea5a9be4aed2.png)

### ❓ pulldown root level

#### 🗹 Left side

![windows-linux-root-level-entries-windows](https://user-images.githubusercontent.com/67791701/225557373-4731d2f9-7be2-469b-a16c-998e4e04cf0f.png)

#### ❓ right side

Yesterday: files on root level shown

![windows-linux-path-pulldown-drives-inux](https://user-images.githubusercontent.com/67791701/225324609-ee2d6ffb-7937-457e-b32d-4696fbf089d4.png)

Now: not shown

![windows-linux-root-level-entries-linux](https://user-images.githubusercontent.com/67791701/225557378-b5f69aa1-a34c-4b51-a4de-0ffcc7714646.png)

Show files on root level or remove the two lines.

### ❌ overwrite dialog

https://user-images.githubusercontent.com/67791701/225563012-305fb4c1-a846-4080-a52a-b1aaef7495b0.mp4

**Comment by grummbeer** [comment]
Windows (VM) > Linux (VM; Ubuntu2204)
---

### ❌ "home" button on right side always refers to "home" of the left side

Always, no matter which direction. 

![windows-linux-home-button-refers-to-left-side](https://user-images.githubusercontent.com/67791701/225566529-629e3cdf-7a1b-4320-947a-dd76758de0e9.png)

**Comment by rustdesk** [comment]
@grummbeer Thanks, related to https://github.com/rustdesk/rustdesk/pull/3560

**Comment by Heap-Hop** [comment]
>"home" button on right side always refers to "home" of the left side

my mistake at https://github.com/rustdesk/rustdesk/pull/3560, just fixed at https://github.com/rustdesk/rustdesk/pull/3680


> remove the two lines.

also adjusted MenuEntryDivider in drop menu. 

Thanks again. @grummbeer

**Comment by grummbeer** [comment]
Thank you! but be aware, will test it in a pedantic way :-)

**Comment by rustdesk** [comment]
👍 
Overwrite is still not working well. @Heap-Hop follow up.

**Comment by rustdesk** [comment]
Of course，i have been thinking about regression test, but it is really hard for our app. I do not think we can do this in current stage.

**Comment by grummbeer** [comment]
please ping me on changes, will try to create a clean test scenario/protocol for manual e2e test.

**Comment by rustdesk** [comment]
I guess except "overwrite dialog", the others has been fixed.

**Comment by grummbeer** [comment]
The overwrite dialog is the topic of this issue report, with still unclear state (for me).

Current state (last test yesterday):

- "Linux > Windows" pass
- "Linux > Android" pass
- "Linux > Linux" failed 
- "Windows > Linux" failed 
- "Windows > Android" not tested so far

Confusing. Need confirmation from some other, is my test case wrong or is it still an issue.

> others has been fixed.

Test for dialog and the others will follow after i am sure that my test env is not broken in any way.

**Comment by rustdesk** [comment]
@Heap-Hop will follow up this, he is still busy workong the other job.

**Comment by rustdesk** [comment]
Wondering if you have interest to dive into this issue? :)

**Comment by grummbeer** [comment]
:-) i knew you will ask … but yes, indeed, i am always interest in seeing new stuff. The component looks complex and it will take time to understand in depth how it works. Let me fix the dialogs first.

**Comment by rustdesk** [comment]
haha

**Comment by olegmm** [comment]
> wait …
> ## Linux > Windows
> ### 🗹 overwrite dialog
> 
> > question"Receive" shows the dialog on every request, but "send" still skips the file silently. Or do i miss something?
> 
> Works … above is **Linux > Linux** … let me double check this.
> ### ❌ Entries on root level not shown
> 
> ![windows-pulldown-double-line](https://user-images.githubusercontent.com/67791701/225304394-c70c55cd-283b-4d3f-9968-a01d47100501.png)
> ### ❌ root level on Windows > Entries/drives not clickable
> 
> ![linux-windows-root-not-clickable](https://user-images.githubusercontent.com/67791701/225304391-4cf0579b-ebeb-4fe2-9d8f-fe8d2a44fe11.png)
> 
>     * "Z" is a network share (if relevant)
> 
>     * "/" as "root" on Windows (instead of "This PC").
> 
> 
> ## Windows > Linux
> ### ❌ overwrite dialog
> 
> Can't test "overwrite" dialog.
> 
> ![windows-linux-root-path](https://user-images.githubusercontent.com/67791701/225321075-580ba83d-5fb3-43d4-8466-02712da9a01d.png)
> 
>     * Also manually set path failing
> 
> 
> Log Windows:
> 
> ```
> flutter: Failed to openDirectory home\: Failed to read dir,timeout
> // manually set path
> flutter: Failed to openDirectory home/stefan\: Failed to read dir,timeout
> ```
> 
>     * "This PC" as "root" on Linux (instead of "/").
> 
> 
> ### 🗹 Entries on root level shown for both sides
> 
> ![windows-linux-path-pulldown-drives](https://user-images.githubusercontent.com/67791701/225324615-a8a046bb-a7ce-40f9-88eb-d135f4a09c2c.png)
> 
> ![windows-linux-path-pulldown-drives-inux](https://user-images.githubusercontent.com/67791701/225324609-ee2d6ffb-7937-457e-b32d-4696fbf089d4.png)



> ## Windows (VM) > Linux (VM; Ubuntu2204)
> ### 🗹 root level / clickable entries
> 
> Yesterday:
> 
> ![windows-linux-root-path](https://user-images.githubusercontent.com/67791701/225321075-580ba83d-5fb3-43d4-8466-02712da9a01d.png)
> 
> Now:
> 
> ![windows-linux-root-level-1-ok](https://user-images.githubusercontent.com/67791701/225555194-dda39212-21ba-49ba-839b-ea5a9be4aed2.png)
> ### ❓ pulldown root level
> #### 🗹 Left side
> 
> ![windows-linux-root-level-entries-windows](https://user-images.githubusercontent.com/67791701/225557373-4731d2f9-7be2-469b-a16c-998e4e04cf0f.png)
> #### ❓ right side
> 
> Yesterday: files on root level shown
> 
> ![windows-linux-path-pulldown-drives-inux](https://user-images.githubusercontent.com/67791701/225324609-ee2d6ffb-7937-457e-b32d-4696fbf089d4.png)
> 
> Now: not shown
> 
> ![windows-linux-root-level-entries-linux](https://user-images.githubusercontent.com/67791701/225557378-b5f69aa1-a34c-4b51-a4de-0ffcc7714646.png)
> 
> Show files on root level or remove the two lines.
> ### ❌ overwrite dialog
> windows-linux-overwrite-dialog.mp4

The nightly build still has this problem ((, I have a lot of clients on linux and I want to use a pro license. But this problem does not allow me to make the transition to rusdesk unfortunately.

Is it possible to resume work on solving this problem ?

**Comment by grummbeer** [comment]
@olegmm are you sure your using a recent nightly?

![3508-not-clickable](https://github.com/rustdesk/rustdesk/assets/67791701/ba63d5e5-111c-44e6-b89a-956b29feb0f8)

The only thing is that pulldown list which is empty on right side (linux). Seems like a regression, i am 75% sure this was fixed by heap-hop ([some comments above](https://github.com/rustdesk/rustdesk/issues/3508#issuecomment-1471722664)) at some time

**Comment by rustdesk** [comment]
@grummbeer Thanks for your confirm.

**Comment by olegmm** [comment]
> @olegmm are you sure your using a recent nightly?
> 

https://github.com/rustdesk/rustdesk/assets/3460830/381f2ef6-8a77-4a08-bead-a90af55166ae



![изображение](https://github.com/rustdesk/rustdesk/assets/3460830/dca1710f-1d7c-4f4e-8e3f-06d8fee261b8)

**Comment by grummbeer** [comment]
mmh, PUTTY.RND is also not transfered, that looks not healthy … but I can not reproduce this. No logs, no idea, looks fine at my side. bad :-(

**Comment by rustdesk** [comment]
@sahilyeole please confirm the remaining issues, and fix.

**Comment by grummbeer** [comment]
Hey, my originally posted issue seems to be fixed? Got no notice and can't find the commits for it at the first glace. Must update me first.

@sahilyeole this report has become a little wild, let me double check what is really still an issue and what not before you investigate.

**Comment by sahilyeole** [comment]
ok sure

**Comment by sahilyeole** [comment]
@grummbeer did u check?

**Comment by grummbeer** [comment]
tl;dr all in all it looks fine :-) The most important is the report from @olegmm see https://github.com/rustdesk/rustdesk/issues/3508#issuecomment-1730099411


- At the moment, i can not test Android as controlling side because of "key mismatch" (Failure::LICENSE_MISMATCH), guess i need a server update? But why the desktops can connect to android … no idea, if you have a quick tip, let me know, if not, don’t care.
- Have no mac to test
- I have tried to test this with the most accuracy and the best knowledge I have, but may miss something.

initial report
---

> 1. sometimes it does not appear at all, files are silently skipped.
> 2. sometimes it appears only in one direction. Forgot which one it was, can't reproduce it at the moment because of "1."
> 3. when selecting multiple files, e.g. "select all", many dialogs are opened, but not as many as files were selected.
> 4. "Do this for all conflicts" does not work, behaves the same as "close this dialog".


- 10 specific files up to 1GB
- Navigate/send/receive from/to different directories (inlcuding symlinks #3501, #3402, #4492 ) 
- bulk copy, one by one copy
- bulk overwrite, one by one overwrite
- bulk delete, one by one delete


### Linux controlling side

- Linux (VM; Ubunut 22.04.1) > Windows 10 (VM) - ✓ PASS
- Linux (VM; Ubunut 22.04.1) > Android 12 (device) - ✓ PASS
- Linux (Tumbleweed) > Linux (VM; Ubunut 22.04.1) - ✓ PASS

### Windows controlling side

- Windows 10 (VM) > Linux (VM; Ubunut 22.04.1) - ✓ PASS
- Windows 10 (VM) > Android 12 (device) - ✓ PASS
- Windows 10 (VM) > Windows 10 (VM) - ✓ PASS

**All issues which are initially reported are solved.**

Seems to be an issue
---

### 1. Flash of exception

After new connection, click some directory on left or right side, a very short flash of this red/yellow flutter execption screen is visible, followed by this logs:

```
Another exception was thrown: A ValueNotifier<_OverlayEntryWidgetState?> was used after being disposed.
Another exception was thrown: Duplicate GlobalKey detected in widget tree.
```

This happens randomly or by a pattern that I can not recongnize.

### 2. Remote "Home" (left side).

Remote "Home" is not "home". Click to "home" (left side) refers to the last opened directory of the previous connection and not to the true user directory. So the button/icon is missleading at the moment.

1. Opening (init with) the last directory when (re)connect makes sense.
2. But click to "home" should always directed to the true "home" directory. As local side does.


Nice to have
---

### 1. bulk delete 

- "Do the same for all" for bulk delete


### 2. Overwrite 

- "Do the same for all conflicts" only if more than one file is selected.


### 3. flag "Canceled" in transfer list

- Skip transfer > "Skipped" - OK
- Do transfer > "Finished" - OK
- Cancel transfer > "Finished" - "Canceled" would be nice


### 4. files on root level for linux

bring back the root level on linux

![windows-linux-path-pulldown-drives-inux](https://user-images.githubusercontent.com/67791701/225324609-ee2d6ffb-7937-457e-b32d-4696fbf089d4.png)


What else
---

### 1. Logs "open direcotry"

Select more than one file, check "do for all" > overwrite. The following logs appear for all, expect the first one.

```
flutter: Failed to openDirectory /storage/emulated/0/Download: Failed to registerReadTask, already have same read job
```

It's not a fault, but some sort of uneccessary log, in meaning of accuracy.


### 2. Transfer list shows full path for windows

When transfering a file from Windows, the full file path is (again) visible in the transfer list. 

I changed this because having the beginning of path hides the essential (the filename).

https://github.com/rustdesk/rustdesk/pull/3498

Maybe reintroduced with the transfer list of the CM?

**Comment by grummbeer** [comment]
tl;dr all in all it looks fine :-) The most important is the report from @olegmm see https://github.com/rustdesk/rustdesk/issues/3508#issuecomment-1730099411


- At the moment, i can not test Android as controlling side because of "key mismatch" (Failure::LICENSE_MISMATCH), guess i need a server update? But why the desktops can connect to android … no idea, if you have a quick tip, let me know, if not, don’t care.
- Have no mac to test
- I have tried to test this with the most accuracy and the best knowledge I have, but may miss something.

initial report
---

> 1. sometimes it does not appear at all, files are silently skipped.
> 2. sometimes it appears only in one direction. Forgot which one it was, can't reproduce it at the moment because of "1."
> 3. when selecting multiple files, e.g. "select all", many dialogs are opened, but not as many as files were selected.
> 4. "Do this for all conflicts" does not work, behaves the same as "close this dialog".


- 10 specific files up to 1GB
- Navigate/send/receive from/to different directories (inlcuding symlinks #3501, #3402, #4492 ) 
- bulk copy, one by one copy
- bulk overwrite, one by one overwrite
- bulk delete, one by one delete


### Linux controlling side

- Linux (VM; Ubunut 22.04.1) > Windows 10 (VM) - ✓ PASS
- Linux (VM; Ubunut 22.04.1) > Android 12 (device) - ✓ PASS
- Linux (Tumbleweed) > Linux (VM; Ubunut 22.04.1) - ✓ PASS

### Windows controlling side

- Windows 10 (VM) > Linux (VM; Ubunut 22.04.1) - ✓ PASS
- Windows 10 (VM) > Android 12 (device) - ✓ PASS
- Windows 10 (VM) > Windows 10 (VM) - ✓ PASS

**All issues which are initially reported are solved.**

Seems to be an issue
---

### 1. Flash of exception

After new connection, click some directory on left or right side, a very short flash of this red/yellow flutter execption screen is visible, followed by this logs:

```
Another exception was thrown: A ValueNotifier<_OverlayEntryWidgetState?> was used after being disposed.
Another exception was thrown: Duplicate GlobalKey detected in widget tree.
```

This happens randomly or by a pattern that I can not recongnize.

### 2. Remote "Home" (left side).

Remote "Home" is not "home". Click to "home" (left side) refers to the last opened directory of the previous connection and not to the true user directory. So the button/icon is missleading at the moment.

1. Opening (init with) the last directory when (re)connect makes sense.
2. But click to "home" should always directed to the true "home" directory. As local side does.


Nice to have
---

### 1. bulk delete 

- "Do the same for all" for bulk delete


### 2. Overwrite 

- "Do the same for all conflicts" only if more than one file is selected.


### 3. flag "Canceled" in transfer list

- Skip transfer > "Skipped" - OK
- Do transfer > "Finished" - OK
- Cancel transfer > "Finished" - "Canceled" would be nice


### 4. files on root level for linux

bring back the root level on linux

![windows-linux-path-pulldown-drives-inux](https://user-images.githubusercontent.com/67791701/225324609-ee2d6ffb-7937-457e-b32d-4696fbf089d4.png)


What else
---

### 1. Logs "open direcotry"

Select more than one file, check "do for all" > overwrite. The following logs appear for all, expect the first one.

```
flutter: Failed to openDirectory /storage/emulated/0/Download: Failed to registerReadTask, already have same read job
```

It's not a fault, but some sort of uneccessary log, in meaning of accuracy.


### 2. Transfer list shows full path for windows

When transfering a file from Windows, the full file path is (again) visible in the transfer list. 

I changed this because having the beginning of path hides the essential (the filename).

https://github.com/rustdesk/rustdesk/pull/3498

Maybe reintroduced with the transfer list of the CM?

**Comment by sahilyeole** [comment]
@grummbeer Thank you so much for the details.
I will test the rest of the environments and check the potential issues.

**Comment by sahilyeole** [comment]
I have tested mac, android on controlling side and mac,android, windows, linux on remote side.
Couldn't find any issue, everything is working well including overwriting.

> Remote "Home" (left right side).
Remote "Home" is not "home". Click to "home" (left right side) refers to the last opened directory of the previous connection and not to the true user directory. So the button/icon is missleading at the moment.
Opening (init with) the last directory when (re)connect makes sense.
But click to "home" should always directed to the true "home" directory. As local side does.

This did not occur in my testing can you be more specific about when it occurred (which environment)? @grummbeer 
For me after reconnecting, the previous connection directory opened and the home was showing the correct home of the remote side.

**Comment by grummbeer** [comment]
All envs.

- connect to linux (right side)
- right side is: /home/stefan - OK, home dir
- navigate to /home/stefan/test-files
- click "Home" > /home/stefan/ - OK, home dir
- navigate to /home/stefan/test-files
- close connection
- reconnect to same linux (right side)
- right side is now: /home/stefan/test-files - OK, last location
- navigate to /home/stefan/Downloads
- click "Home" > /home/stefan/test-files - KO, last location from previous connection.

**Comment by sahilyeole** [comment]
Reproduced, Thanks.

I could not find any direct implementation for the remote home directory yet but since the home path is correct in the first connection we can just store it according to the session ID and fetch whenever needed.
I'll do it soon.

**Comment by grummbeer** [comment]
finally. Thanks @sahilyeole! :+1:

**Comment by olegmm** [comment]
note
if someone encounters a problem with displaying directory listings for the second user in the system

your initial user uid  must be => 1000

in that case you normal can see it gdm3 login user list and can see file listing in rusdesk file-transfer window
